### PR TITLE
Refactor and test journal pruning class

### PR DESCRIPTION
**Pull request by AlexandraRoatis**
## Description

Please include a brief summary of the change that this pull request proposes. Include any relevant motivation and context. List any dependencies required for this change.

- made the `JournalPruneDataSource.java` independent of the block and block header classes
- improved and corrected synchronization
- removed locks from `stateDatabase` since the only class accessing it is the `JournalPruneDataSource`
- further unit tests for the journal prune functionality

Related to issues #54 and #298.

## Note: this PR builds on top of #478.  

The PR in #478 should be reviewed and merged first. After which this PR can be rebased to `master-pre-merge` branch.

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [x] New feature.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

Please describe the tests you used to validate this pull request. Provide any relevant details for test configurations as well as any instructions to reproduce these results.

- full test suite and unit tests submitted as part of this PR

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [ ] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java [hunk 1]
```diff
@@ -157,51 +213,64 @@ private Ref decRef(ByteArrayWrapper keyW) {
...
-    public synchronized void storeBlockChanges(BH header) {
-        if (!enabled) {
-            return;
+    public void storeBlockChanges(byte[] blockHash, long blockNumber) {
+        lock.writeLock().lock();
+
+        try {
+            if (!enabled) {
+                return;
+            }
```
Comment (AionJayT) [inline modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java]: should we check enable before the writelock?

File: modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java [hunk 2]
```diff
@@ -85,58 +90,109 @@ public JournalPruneDataSource(IByteArrayKeyValueDatabase src) {
     }
 
     public void setPruneEnabled(boolean e) {
+        lock.writeLock().lock();
+
         enabled = e;
+
```
Comment (AionJayT) [inline modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java]: use atomicboolean for the variable - enabled, then don't need lock in this method also check enabled before the other methods before writelock.

File: modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java [hunk 3]
```diff
@@ -222,29 +291,54 @@ private synchronized void rollback(BH header) {
...
     }
 
     public Optional<byte[]> get(byte[] key) {
-        return src.get(key);
+        lock.readLock().lock();
+        try {
+            return src.get(key);
+        } catch (Exception e) {
+            throw e;
+        } finally {
```
Comment (AionJayT) [inline modMcf/src/org/aion/mcf/trie/JournalPruneDataSource.java]: for the following throws(including this) , add an error log before throw if no outside classed catch these exceptions.

**Comment by AlexandraRoatis** [comment]
rebased the branch to include changed to #478

**Comment by qoire** [comment]
General question about the reference counts: Is it possible to remove the DB read that happens when we initialize a ref in ``incRef()``?

We only seem to use it in removing fork blocks, so I believe we can do a check if absolutely necessary in that scenario but otherwise we can ignore it.

**Pull request merged**